pkg/kubectl/cmd/set/set_env.go

```diff
@@ -203,18 +221,75 @@ func (o *EnvOptions) Complete(f cmdutil.Factory, cmd *cobra.Command, args []stri
 	o.Output = cmdutil.GetFlagString(cmd, "output")
 	o.From = cmdutil.GetFlagString(cmd, "from")
 	o.Prefix = cmdutil.GetFlagString(cmd, "prefix")
+	o.Keys = cmdutil.GetFlagStringSlice(cmd, "keys")
 	o.DryRun = cmdutil.GetDryRunFlag(cmd)
 
-	o.EnvArgs = envArgs
-	o.Resources = resources
 	o.Cmd = cmd
 
 	o.ShortOutput = cmdutil.GetFlagString(cmd, "output") == "name"
 
+	if len(o.From) == 0 {
+		return nil
+	}
+	cmdNamespace, enforceNamespace, err := f.DefaultNamespace()
+	if err != nil {
+		return err
+	}
+	b := f.NewBuilder().
+		Internal().
+		LocalParam(o.Local).
+		ContinueOnError().
+		NamespaceParam(cmdNamespace).DefaultNamespace().
+		FilenameParam(enforceNamespace, &o.FilenameOptions).
+		Flatten()
+
+	if !o.Local {
+		b = b.
+			LabelSelectorParam(o.Selector).
+			ResourceTypeOrNameArgs(o.All, o.From).
+			Latest()
+	}
+
+	infos, err := b.Do().Infos()
```

Comment:
This operation does not belong in `Complete`. `Complete` should fill in the necessary bits of the `EnvOptions` struct so that they are then validated and an actual operation is performed in `Run`. This should be moved to `Run`.